klog.go

```diff
@@ -748,6 +748,53 @@ func (l *loggingT) printWithFileLine(s severity, logr logr.InfoLogger, file stri
 	l.output(s, logr, buf, file, line, alsoToStderr)
 }
 
+// printS if loggr is specified, no need to output with logging module. If
+// err arguments is specified, will call logr.Error, or output to errorLog severity
+func (l *loggingT) printS(err error, loggr logr.Logger, msg string, keysAndValues ...interface{}) {
+	if loggr != nil {
+		if err != nil {
+			loggr.Error(err, msg, keysAndValues)
+		} else {
+			loggr.Info(msg, keysAndValues)
+		}
+		return
+	}
+	b := &bytes.Buffer{}
+	b.WriteString(fmt.Sprintf("%q", msg))
+	if err != nil {
+		b.WriteByte(' ')
+		b.WriteString(fmt.Sprintf("err=%q", err.Error()))
+	}
+	kvListFormat(b, keysAndValues...)
+	var s severity
+	if err == nil {
+		s = infoLog
+	} else {
+		s = errorLog
+	}
+	l.printDepth(s, logging.logr, 1, b)
+}
+
+const missingValue = "(MISSING)"
+
+func kvListFormat(b *bytes.Buffer, keysAndValues ...interface{}) {
+	for i := 0; i < len(keysAndValues); i += 2 {
+		var v interface{}
+		k := keysAndValues[i]
+		if i+1 < len(keysAndValues) {
+			v = keysAndValues[i+1]
+		} else {
+			v = missingValue
+		}
+		b.WriteByte(' ')
+		if _, ok := v.(fmt.Stringer); ok {
+			b.WriteString(fmt.Sprintf("%s=%q", k, v))
+		} else {
+			b.WriteString(fmt.Sprintf("%s=%#v", k, v))
```

Comment:
My question was with keys, not values.

%s key produces all sorts of weird strings for non-string types, but I guess that's fine as long as "not worse than" before.

%#v is fine for value, as long as the goal is "not worse than", but it does make more or less unparseable logs.